### DISCUSSION TITLE
Unsigned apps on M1 Macs

### DISCUSSION THREAD

**Discussion by alexreg** [Homebrew/discussions#3088] [Casks] [answered]
There's a problem with unsigned apps on the new M1 Macs (at least, I believe it's an M1 thing rather than a new version of macOS) — namely, it's impossible to run unsigned apps without self-signing them first. Whereas before, unsigned apps seem to have been treated the same as apps signed by unknown developers, now you won't even get a prompt in the *Security* pane of System Preferences to "Allow" (add an exception). Consequently, after installing a cask with an unsigned app, it will refuse to open. e.g., from Finder, you get a message like, `The application “Foo.app” can’t be opened.` with an error message about security in the Console.

The obvious solution seems to be to self-sign all apps that Homebrew Cask installs (at least on M1 Macs). This is typically done via the command line with

`sudo codesign --force --deep -s - Foo.app`

Unsurprisingly, it requires admin privileges. Alternatively, a post-install not could just be displayed to the user to tell them to do this, but this is obviously less than ideal in light of automatically updated casks and their apps.

Any thoughts about how Homebrew should deal with this situation?

**Comment by SMillerDev** [comment]
Does this solution https://docs.brew.sh/FAQ#why-cant-i-open-a-mac-app-from-an-unidentified-developer no longer work?

**Reply by alexreg** [reply]
Yeah, I don't get that message any more, and the right-click & open solution doesn't help either.

**Reply by alexreg** [reply]
Two casks this happens with are Radarr and Prowlarr.

**Reply by neutric** [reply]
What about

> Alternatively, you may provide the [--no-quarantine flag](https://github.com/Homebrew/homebrew-cask/blob/HEAD/USAGE.md#options) at install time to not add this feature to a specific app.

?

**Reply by alexreg** [reply]
@neutric Nope, no luck I'm afraid.

**Reply by badonyx** [reply]
On Apple Silicon, `--no-quarantine` flag works for some casks but not for others. I did some investigation on this, here are my findings.

Per [Apple](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-universal-apps-release-notes/#Code-Signing) (see also [Apparency](https://www.mothersruin.com/software/Apparency/faq.html#signature-status)) there is a distinction between apps that are:

- Built for Intel, or built for Apple Silicon
- Totally unsigned, or signed ad-hoc by macOS linker

> New in macOS 11 on Macs with Apple silicon, and starting in macOS Big Sur 11 beta 6, the operating system enforces that any executable must be signed before it’s allowed to run. There isn’t a specific identity requirement for this signature: a simple ad-hoc signature is sufficient. This new behavior doesn’t change the long-established policy that our users and developers can run arbitrary code on their Macs, and is designed to simplify the execution policies on Macs with Apple silicon and enable the system to better detect code modifications. This new policy doesn’t apply to translated x86 binaries running under Rosetta 2, nor does it apply to macOS 11 running on Intel-based platforms.

So this means, and I have confirmed, that:

- An app WILL run on Apple Silicon devices with quarantine attributes removed if it is built for Intel devices, OR if it is built for Apple Silicon devices and signed ad-hoc by linker. (We can call this category "unidentified developer" and the homebrew FAQ still applies.)
  - Known casks: chromium, chrome-driver, bloodhound, mark-text, copyq
- An app WILL NOT run on Apple Silicon devices if it is built for Apple Silicon and totally unsigned.
  - Known casks: radarr, prowlarr

This last category is the most problematic one, but it also appears to be an upstream issue if the developers are distributing apps that have never actually worked for the target platform. As such I do not think it makes sense to add a signing mechanism for those casks, rather the Apple Silicon builds should be unlisted until the developers produce working builds.

**Reply by alexreg** [reply]
@badonyx Basically, the *rr maintainers don't want to buy a signing certificate, and attach any of their identities to one, I've been told.

**Reply by badonyx** [reply]
A signing certificate is not needed, as you can see from the other category of "unsigned" apps that are runnable (after removing quarantine attributes). To give more context from the next paragraph in the apple documentation:

> To reduce the impact on existing development workflows, starting with Xcode 12 beta 4, the toolchain will now automatically sign your executables whenever you build from Xcode, or use command-line utilities such as clang(1) or ld(1). Since this signing mechanism generates signatures directly at link time, and doesn’t cover any resource other than the executable, it is expected to be faster than a traditional codesign(1) invocation. If you use a custom workflow involving tools that modify a binary after linking (e.g. strip or install_name_tool) you might need to manually call codesign(1) as an additional build phase to properly ad-hoc sign your binary. These new signatures are not bound to the specific machine that was used to build the executable, they can be verified on any other system and will be sufficient to comply with the new default code signing requirement on Macs with Apple silicon. However, given that these signatures do not bear any valid identity, binaries signed this way cannot pass through Gatekeeper.

**Reply by alexreg** [reply]
@badonyx Okay, so that still incurs the same problem: macOS won't let them run on ARM chips?

**Reply by badonyx** [reply]
It can run (unquarantined) if they rework their build pipeline to incorporate the ad-hoc signature from the apple toolchain.

**Reply by alexreg** [reply]
Ah okay, that's what I thought. Looks like they have no desire to do so, judging by the responses I've got from the maintainers.

**Reply by badonyx** [reply]
The unsigned intel apps should still work (with rosetta) so until they decide to really support Apple Silicon you can simply remove the arm builds from the casks (and add the rosetta caveat).

**Reply by badonyx** [reply]
Submitted PRs for this.

- https://github.com/Homebrew/brew/pull/16243
- https://github.com/Homebrew/homebrew-cask/pull/160619

**Comment by lougreenwood** [comment]
I also have a similar issue. It seems that self-signed apps won't auto-launch.

Or possibly more specifically - self-signed apps which are installed by another user who is admin and the current user is not admin.

I need to investigate this some more, but it's super annoying!

**Comment by shemeshg** [comment]
# is it still ok?

Supposed the app is failed with
```
  * line 5, col 2: Signature verification failed:
    /private/tmp/d20231117-26929-727bs7/pass-simple.app: rejected

    macOS on ARM requires software to be signed.
    Please contact the upstream developer to let them know they should sign and notarize their software.
Error: 3 problems in 1 cask detected.
```

Maybe the request to notarize for open source developer is too much (They will not pay the fee)? 


## asking to sign and not notarize actually  increase `homebrew` security

asking developers to 
```
codesign --deep --timestamp --options runtime -s "Apple Development" application-name.app
```
might help auto approve upgraded versions from the same developer.

* Maybe homebrew can use `codesign -dv --verbose=4  app-name.app` for verification instead ?

however mandatory `notarize` might be too much...

## According to ChatGpt

If your app is signed but not notarized, it will still open on another Mac. However, users might encounter a security warning.
The error message you received indicates that the app failed signature verification. While it will run, macOS on ARM requires signed software. Notarization is recommended for enhanced security.

complete log:
```
macos@macoss-Mac-mini Casks % brew audit --new-cask quicktest
==> Downloading and extracting artifacts
==> Downloading https://downloads.sourceforge.net/pass-simple/quicktest.dmg
==> Downloading from https://master.dl.sourceforge.net/project/pass-simple/quick
######################################################################### 100.0%
==> Downloading https://downloads.sourceforge.net/pass-simple/quicktest.dmg
Already downloaded: /Users/macos/Library/Caches/Homebrew/downloads/404a43a9727e0b574c9b52bdb3e9b58aac8f41f3447cea2107bd2b031336a2c5--quicktest.dmg
audit for quicktest: failed
 - Use `sha256 :no_check` when URL is unversioned.
 - Download is hosted on SourceForge, please add a livecheck. See https://docs.brew.sh/Cask-Cookbook#stanza-livecheck
 - Signature verification failed:
/private/tmp/d20231117-26929-727bs7/pass-simple.app: rejected

macOS on ARM requires software to be signed.
Please contact the upstream developer to let them know they should sign and notarize their software.
shemeshg/tap/quicktest
  * Use `sha256 :no_check` when URL is unversioned.
  * line 5, col 2: Download is hosted on SourceForge, please add a livecheck. See https://docs.brew.sh/Cask-Cookbook#stanza-livecheck
  * line 5, col 2: Signature verification failed:
    /private/tmp/d20231117-26929-727bs7/pass-simple.app: rejected

    macOS on ARM requires software to be signed.
    Please contact the upstream developer to let them know they should sign and notarize their software.
Error: 3 problems in 1 cask detected.
```

**Reply by SMillerDev** [reply]
This doesn't actually work. The certificate is paid as well as getting a binary notarised. Homebrew will not change the detection because otherwise apps will not work on any Mac released in the last 3 years .

**Reply by shemeshg** [reply]
Thanks for your response.

To summarize that:

* Unfortunately, this means that even if an open-source developer contributes to a non-profit project, 
    they might not be eligible for certain privileges unless they have a personal  Developer Program Membership.

* Unless "Apple Developer Program Membership Fee Waivers" (non profit organization) is wiling to submit on his behalf.

**Reply by alexreg** [reply]
I still think the obvious solution is for Homebrew to come with its own code-signing certificate and use that to (re-) self-sign the app.

**Reply by shemeshg** [reply]
I think so too @alexreg, 
I guess you mean `homebrew` self signed.

## or another other option (For not notarised).
* Developer will
`codesign --deep --timestamp --options runtime -s "Apple Development" application-name.app` 

* homebrew will look in to the `codesign -dv --verbose=4  app-name.app`
## For new app :
* it will extract the developer id ( `username@me.com` and gpg stub,
*  If approved, store it in JSON file in it's repository
## for version bump
* when ever version bump received it will compare against `username@me.com`,gpg stub and cask name, and if match automatically approve.

Proposed Json/Yaml
```
- Cask name: whatever.rb 
  Developer Id: abc@me AFTYAS981234
- Cask name: gogo.rb 
  Developer Id: asdfasdf@me UIOAS981234

```

**Reply by SMillerDev** [reply]
> Unfortunately, this means that even if an open-source developer contributes to a non-profit project, they might not be eligible for certain privileges unless they have a personal Developer Program Membership.

Not really, all it means that if nobody makes sure there is a paid Apple Developer Account you can't have a signed public application.

> Unless "Apple Developer Program Membership Fee Waivers" (non profit organization) is wiling to submit on his behalf.

That's one way to obtain signing without paying on a personal level.

> I still think the obvious solution is for Homebrew to come with its own code-signing certificate and use that to (re-) self-sign the app.

That will not happen since Homebrew can never validate that the precompiled binary is legitimate and non-malicious. And it will not pretend it can by signing everything.

**Reply by alexreg** [reply]
@SMillerDev Ah I see. Is there any possibility of generic post-install hooks for formulas/casks though? So maybe users could automate this ourselves if needed? I have something already in place, but it's a little hacky.

**Reply by shemeshg** [reply]
@SMillerDev 
1. Is there a scheduled due date or a cutoff point when non-notarized software on Homebrew Casks will be removed? Additionally, 
2. until which date will updates for non-notarized software continue to be supported for existing Homebrew Casks?

**Reply by SMillerDev** [reply]
> Ah I see. Is there any possibility of generic post-install hooks for formulas/casks though? So maybe users could automate this ourselves if needed? I have something already in place, but it's a little hacky.

Formulae are automatically ad-hoc signed already, since there Homebrew builds it so we can't ask upstream to sign it.

As for a general solution, you can always make a pull request.

> Is there a scheduled due date or a cutoff point when non-notarized software on Homebrew Casks will be removed? 

I don't have any plans to remove existing software, but new software without signing is not allowed because the user experience is so bad.
> Additionally,
until which date will updates for non-notarized software continue to be supported for existing Homebrew Casks?

In the main tap, I have no plans to extend the signing check to existing casks

**Reply by alexreg** [reply]
@SMillerDev Thanks for the clarification. How would you like to see a general solution designed, if I do take that root? Is a generic post-install hook (for casks only or formulas & casks) a fair approach to you?

**Reply by shemeshg** [reply]
@SMillerDev 

If I understand you correctly, the recommended way to distribute non-notarized executables is to use formula instead of cask. However, I could not find any clear guidance or examples on how to **formula** gui applications for Homebrew. This applies to most used frameworks 
1. `electron`+`javascript` 
    Consider for example, Electron deployment tool, that today used for automatically  produce DMG, snap, deb, NSIS will need to know how to produce `homrbew formula` also 
3. `cmake + qt6 + qml + c++` frameworks.
4. xcode swiftUi gui native applications

that to say this issue is specific to gui applications, as there are plenty of resources and documentation for libraries and non-gui.

* is it planned to publish best practices for that?
* for this is a **huge** change, is it gone be communicated in your blog, including list to example formulas best practices?
* How about soft due dates (with no cutoff), for helping developers to make the transition, and dedicated pinned discussion thread.
* Adjusting formula to handle desktop shortcut, and linking etc... (integrating functionality and tags exists in cask, so it will be available in formulas)

**Reply by SMillerDev** [reply, marked as answer]
> Thanks for the clarification. How would you like to see a general solution designed, if I do take that root? Is a generic post-install hook (for casks only or formulas & casks) a fair approach to you?

I'm not sure, but that's probably something that can be worked out in a review 

> If I understand you correctly, the recommended way to distribute non-notarized executables is to use formula instead of cask. However, I could not find any clear guidance or examples on how to formula gui applications for Homebrew.

It is not, the recommended way to distribute GUI applications is as signed casks.

**Reply by alexreg** [reply]
@SMillerDev Okay, sure. I was thinking of generic pre-post install scripts for all formulas/casks, to be searched for at a fixed set of paths. Could you kindly point me to the right place in the codebase I could look at, when it comes to running these scripts pre/post-installation of a formula/cask?

**Reply by SMillerDev** [reply]
No such place currently exists, so it's totally up to you

**Reply by alexreg** [reply]
> No such place currently exists, so it's totally up to you

Surely there's an install function/method that I should look at at least...

**Reply by shemeshg** [reply]
It could be that initiating work without a detailed technical description of the proposed solution would be premature. 
Additionally, the term ‘hooks’ seems to have caused confusion, as it could refer to 

- GitHub server hooks, 
- - Ruby-level callbacks, or 
- - new tags within the Cask implementation.

Maybe @SMillerDev can briefly describe expected solution? 

Also now that we know( @badonyx) Apple original intensions we just and only just binary completeness,
Maybe the solution is replacing  one line of code from current new cask validation, to validating existence of app bundle signature
(notarise by apple or not)

**Reply by SMillerDev** [reply]
I can not describe an expected solution since I do not expect any solution, if people think they can come up with a system to sign software locally they are totally free to design and suggest an implementation.

**Reply by ferdnyc** [reply]
@shemeshg 
> Also now that we know( @badonyx) Apple original intensions we just and only just binary completeness,

...That sounds like a mischaracterization, TBH. The _function_ of code signing is to verify that an app's contents have not been **modified** since the app was signed. (Apple actually [puts it as](https://support.apple.com/en-gb/guide/security/sec3ad8e6e53/web), "verification of this signature proves to the user that a developer’s software hasn’t been tampered with since the developer built and signed it.") But that's because code signing is only part of their overall application security framework.

Apple provides the ability to bypass signature identity checks and notarization requirements as _developer_ features, so that software developers and advanced users who know what they're doing are able to run code developed and built locally without having to put it through the full application security process.

But the intent has always been that those options aren't for general use, and that applications built that way shouldn't be distributed to end users. That's what Apple means, in the docs @badonyx quoted, when they say, "However, given that these signatures do not bear any valid identity, binaries signed this way cannot pass through Gatekeeper." Gatekeeper, not code signing, is the primary defense against potentially malicious code in macOS. Apple has a whole other document about it, [Gatekeeper and runtime protection in macOS](https://support.apple.com/en-gb/guide/security/sec5599b66df/web).

> macOS includes a security technology called Gatekeeper, which is designed to help ensure that only trusted software runs on a user’s Mac. When a user downloads and opens an app, a plug-in or an installer package from outside the App Store, Gatekeeper verifies that the software is from an identified developer, is notarised by Apple to be free of known malicious content and hasn’t been altered. Gatekeeper also requests user approval before opening downloaded software for the first time to make sure the user hasn’t been tricked into running executable code they believed to simply be a data file.

Apple's _intent_ is clearly that Homebrew-installed packages would be subject to Gatekeeper verification, same as any other downloaded software. And even if it's possible for Homebrew to make installed apps appear as trusted to the OS, it probably _shouldn't_. That would be accepting a lot of responsibility for ensuring that the software is genuinely benign and safe to run.

Just like, whenever they provide instructions for overriding application security for a specific app, Apple [always say](https://support.apple.com/en-us/HT202491) (emphasis added):

> **If you're certain that the app is from a trustworthy source and hasn't been tampered with**, you might be able to temporarily override your Mac security settings to open it

Can the Homebrew team genuinely say that about any given Cask of Radarr, Prowlarr, or... well, really _any_ package made available via `brew install`? Homebrew **itself** may be a trustworthy source, but conveying that trust by proxy to the developers of all of the packages in the collection seems a bit cavalier.

**Reply by SMillerDev** [reply]
> Can the Homebrew team genuinely say that about any given Cask of Radarr, Prowlarr, or... well, really any package made available via brew install?

For casks we can not and will not. For formulae however we build them so we are confident about the source of the install.

**Reply by SylvanFranklin** [reply]
Any progress on this? I'm also trying to create a tap for an unsigned binary, and I'm having trouble finding documentation regarding creating a gui based application outside of cask. 

> @SMillerDev
> 
> If I understand you correctly, the recommended way to distribute non-notarized executables is to use formula instead of cask. However, I could not find any clear guidance or examples on how to **formula** gui applications for Homebrew. This applies to most used frameworks
> 

I'm wondering the same thing.

**Reply by shemeshg** [reply]
Cask Tap is a standard repository, like this one:

https://github.com/shemeshg/homebrew-tap/blob/master/Casks/pass-simple.rb

The differences are:

1. Unlike formulas, you’re expected to deploy your binary (using cmake’s cpack, qt-deploy, or manually copying files...).

2. Users should use `xattr - c /Applications/your app` to disable authorization check.

Specific authorization won’t work, and users will need to run the app through the terminal with the following command:

pass-simple=‘nohup /Applications/pass-simple.app/Contents/MacOS/pass-simple > /dev/null 2>&1 &’

This uses terminal authorization on behalf of your app.

* theoretically you can create formula for gui app, and `cmake install` will do it alll for you, but I haven't seen example for that.
Post example here if you find one. (preferred simple `hello world` qt qml one, notice you will have to implement the uninstall and zap equivalent)

**Reply by SMillerDev** [reply]
> I'm having trouble finding documentation regarding creating a gui based application outside of cask.

Formula can build anything, including GUI apps, but they're not very convenient for GUI based applications because they won't show up in the usual places in the OS. (`/Aplications` for example).
> If I understand you correctly, the recommended way to distribute non-notarized executables is to use formula instead of cask.

The recommendation is to notarise your executables. If you can't do that, building from source in a formula is an alternative option.
> Users should use xattr - c /Applications/your app to disable authorization check.

Users should most certainly not do that. There is a reason the macOS gatekeeper exists, circumventing it like this it harms the security of the users system.

**Reply by SylvanFranklin** [reply]
My current workaround is installing the app bundle from the dmg, and putting --no-quarantine in my install instructions along with a warning about why. Not ideal but it's functional.
